### DISCUSSION TITLE
Any sample Docker scripts?

### DISCUSSION THREAD

**Discussion by jgfoster** [eclipse-theia/theia#11760] [Q&A]
I'm having trouble creating a Docker image for Theia and would appreciate help.

[Theia-apps](https://github.com/theia-ide/theia-apps/tree/master/theia-docker) repository is archived (and doesn't work). [Theia-blueprint](https://github.com/eclipse-theia/theia-blueprint/blob/master/Dockerfile) has a `Dockerfile` that builds and tests Python 3.6 from sources—but has no Theia in it! 

I've created my own based on the documentation I can find (see below), but when I run the image and connect to it (`curl -v localhost:3000`), I get an empty reply (but it does connect, so I believe the port is correctly mapped). When I do the `curl` command from within the Docker container, it replies with the expected file. When I run an Nginx Docker container it works fine, so this seems to be some combination of Theia and Docker that doesn't work together (each works fine separately).

Any ideas?

```
FROM ubuntu:22.04

# -----------------------------------------------------------------------------
# Prerequisites: https://github.com/eclipse-theia/theia/blob/master/doc/Developing.md#prerequisites
RUN apt-get update && \
    apt-get install -y curl && \
    curl -fsSL https://deb.nodesource.com/setup_16.x | bash -  && \
    curl -fsSL https://dl.yarnpkg.com/debian/pubkey.gpg | apt-key add - && \
    echo 'deb https://dl.yarnpkg.com/debian/ stable main' | tee /etc/apt/sources.list.d/yarn.list && \
    apt-get update && \
    apt-get install -y nodejs yarn build-essential git libsecret-1-dev libx11-dev libxkbfile-dev pkg-config

# -----------------------------------------------------------------------------
# Install and build Theia
RUN git clone https://github.com/eclipse-theia/theia /opt/theia
WORKDIR /opt/theia
ADD package.json .
RUN NODE_OPTIONS="--max_old_space_size=2048" yarn
# ADD settings.json /etc/theia/

EXPOSE 3000
CMD yarn start

# -----------------------------------------------------------------------------
# Build and run commands

# docker build -t theia-java .
# docker run --rm -it -p 3000:3000 theia-java
# curl -v localhost:3000
```

**Comment by msujew** [comment]
Hey @jgfoster,

so this is a relatively simple docker image that we use to create a theia browser application:

```dockerfile
FROM node:16

RUN apt-get update && apt-get install -y libsecret-1-dev

WORKDIR /usr/src/app

RUN yarn global add lerna

COPY theia-app ./theia-app
COPY theia-extension ./theia-extension
COPY package.json .

COPY lerna.json .
RUN lerna bootstrap

RUN mkdir /workspace
WORKDIR /usr/src/app/theia-app
CMD yarn theia start /workspace --hostname 0.0.0.0 --plugins=local-dir:./plugins
```

It's not the most well optimized - the guys over at Ericsson probably have something better - but it works for us.

**Reply by jgfoster** [reply]
It is helpful to know that someone has done this successfully. When I try to run this I get errors with the `COPY` commands. Can you share what is in them?

**Reply by msujew** [reply]
You can probably simplify this down even more to just the following. Note that this expects a setup similar to what the [theia yeoman generator](https://github.com/eclipse-theia/generator-theia-extension/) does.

```dockerfile
FROM node:16

RUN apt-get update && apt-get install -y libsecret-1-dev

WORKDIR /usr/src/app

COPY . .
RUN yarn
RUN yarn build

RUN mkdir /workspace
# This should be the directory of your browser application
WORKDIR /usr/src/app/browser-app
CMD yarn theia start /workspace --hostname 0.0.0.0 --plugins=local-dir:./plugins
```

**Reply by jgfoster** [reply]
Thanks very much for the quick reply. Unfortunately, when I follow the instructions for the [theia yeoman generator](https://github.com/eclipse-theia/generator-theia-extension/) I get a number of errors, ending with `Error: webpack exited with an unexpected code: 1`. So that didn't get me very far. I'm interested that your approach relies on a local environment that is built outside the Dockerfile script. By contrast, I'm trying to keep everything inside the script. In any case, thanks for the comments.
